Ufunc examples docs (#464)
xgcm/xgcm: doc/grid_ufuncs.rst

.. _grid_ufuncs:

Grid Ufuncs
-----------

Concept of a Grid Ufunc
~~~~~~~~~~~~~~~~~~~~~~~

In short, a "grid ufunc" is a generalisation of a `numpy generalized universal function`_ to include the xGCM Axes and Positions of input and output variables.

If you are not already familiar with numpy generalised universal functions (hereon referred to as "numpy ufuncs"),
then here is a quick primer.

.. dropdown:: **Primer on numpy generalized universal functions**

>     Content about numpy...

Reviewer: Do we want to fill this in before merging? Or otherwise just link out to the numpy docs.

Author: Oh oops - I was originally going to write out my own explanation here but actually I think a link to the numpy docs should be sufficient.

We tell a function about the axes information through a ``signature``.
For example, imagine we have a function which accepts data located at the center grid positions and returns
data located at the left-hand grid positions, on the same axis.
The signature for this function would look like:

``"(ax1:center)->(ax1:left)"``.

> You will also need to understand the `concept of "core dims"`_ used in ``xarray.apply_ufunc``.

Reviewer: This might make sense as a prerequisite reading at the top of the page.

Grid ufuncs allow us to:

- Avoid mistakes by stating that functions are only valid for data on specific grid positions,
- Neatly promote numpy functions to grid-aware xarray functions,
- Conveniently apply boundary conditions and grid topologies (see :ref:`Boundaries and Padding`),
- Boost performance relative to chaining many `Grid.diff` or `Grid.interp` operations,
- Immediately parallelize our operations with dask (see :ref:`Parallelizing with Dask`).

.. _numpy generalized universal function: https://numpy.org/doc/stable/reference/c-api/generalized-ufuncs.html
.. _concept of "core dims": https://xarray.pydata.org/en/stable/generated/xarray.apply_ufunc.html

The ``signature``
~~~~~~~~~~~~~~~~~

The "signature" of a grid ufunc is how we tell it which input and output variables should be on which axis positions.
A simple example would be
``"(ax1:center)->(ax1:left)"``

The signature has two parts, one for the input variables, and one for the output variables.
The output variables live on the right of the arrow (``->``).

> There needs to be one bracketed entry for each variable,

Reviewer: ```suggestion
There needs to be one entry in parentheses for each variable,
```

so in this case the signature tells use that the function accepts one input data variable and returns one output data variable.
(Functions which accept a data variable in the form of a keyword-only argument are not supported.)

For each variable, the signature tells us the ``xgcm.Axis`` positions we require that variable to have,
both before and after our grid ufunc is applied.
This information is encoded in the format ``axis_name:position``.
Each variable can be operated on along multiple axes, which are separated by a comma, e.g. ``(ax1:left, ax2:right)``.

The axis names used in the signature are dummy names: they do not have to be the same as the axis names used in your ``Grid`` object.

> This allows you to write a grid ufunc that can accept axis with any name.

Reviewer: ```suggestion
This allows you to write a grid ufunc that can accept axes with any name.
```

   * - ``diff_left_to_center(a)``
     - ``"(X:left)->(X:center)"``
     - Forward difference
   * - ``interp_left_to_center(a)``
     - ``"(X:left)->(X:center)"``
     - Forward interpolation
   * - ``diff_center_to_center(a)``
     - ``"(X:center)->(X:center)"``
     - Second order central difference
   * - ``mean_depth(w)``
     - ``"(depth:center)->()"``
     - Reduction
   * - ``inner_product_left_right(a, b)``
     - ``"(X:left),(X:right)->()"``
     - Reduction of two variables
   * - ``u_grid_vorticity(u, v)``
     - ``"(lon:left,lat:center),``
       ``(lon:center,lat:left)``
       ``->(lon:left,lat:left)"``
     - Complex example calculating vorticity on an Arakawa U-grid in 2D

.. note::

    Remember the axis names in the signature are dummy names - in the example above you could apply ``mean_depth`` along
    an axis not called ``"depth"`` if you wish.

    The ``axis`` argument is the one which must correspond to an ``xgcm.Axis`` of the grid.
    Therefore applying a grid ufunc with signature ``"(X:center)->()"`` or ``"(depth:center)->()"`` along ``axis='X'`` will
    yield identical results in both cases.

Defining New Grid Ufuncs
~~~~~~~~~~~~~~~~~~~~~~~~

> Lets imagine we have a numpy function which does forward differencing along one dimension, with an implicit periodic boundary condition.

Reviewer: ```suggestion
Let's imagine we have a numpy function which does forward differencing along one dimension, with an implicit periodic boundary condition.
```

.. ipython:: python

    import numpy as np

    def diff_forward(a):

>         return a - np.roll(a, -1, axis=-1)

Reviewer: I find the use of roll here a little confusing, because it brings up boundary conditions. We haven't really introduced how boundary conditions will be handled yet.

Reviewer: Actually, it seems fine after reading the next section.

Now when we run our decorated function `interp_center_to_left`, xgcm will automatically add an extra element to the left hand side for us,
before applying the operation in the function we decorated.

.. ipython:: python

    # Create new test data with same coordinates but linearly-spaced data
    da = da.copy(data=arr)

    interp_center_to_left(grid, da, axis=[["X"]])

Here a periodic boundary condition has been used as the default, but we can choose other boundary conditions using the ``boundary`` kwarg:

.. ipython:: python

    @as_grid_ufunc(
        signature="(X:center)->(X:left)",
        boundary_width={"X": (1, 0)},
        boundary="fill",
        fill_value=0,
    )
    def interp_center_to_left_fill_with_zeros(a):
        return interp(a)

    interp_center_to_left_fill_with_zeros(
        grid, da, axis=[["X"]], boundary="fill", fill_value=0
    )

We can also choose a different default boundary condition at decorator definition time,
and then override it at function call time if we prefer.

.. ipython:: python
    :okexcept:

    interp_center_to_left(grid, da, axis=[["X"]], boundary="fill", fill_value=np.NaN)

- Link to more specific docs?

> - Link to more complex examples?

Reviewer: Need to be addressed.

Metrics
~~~~~~~

.. note::

>     Supplying metrics directly to grid ufuncs is not yet implemented, but will be soon!

Reviewer: ```suggestion
    Automatically supplying metrics directly to grid ufuncs is not yet implemented, but will be soon!
    For now, if you need a metric in your grid ufunc, simply include it as an input and pass it explicitly. 
```
